cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+		return nodeInfo, nil
+	}
+	if data.deletedNodeInfos[name] || data.baseData == nil {
+		return nil, fmt.Errorf("node not found")
+	}
+	return data.baseData.GetNodeInfo(name)
+}
+
+func (data *internalDeltaSnapshotData) GetNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	if data == nil {
+		return nil
+	}
+	if data.completeNodeInfoList == nil {
+		data.buildNodeInfoList()
+	}
+	return data.completeNodeInfoList
+}
+
+// buildNodeInfoList contains costly copying throughout the struct chain. Use wisely.
+func (data *internalDeltaSnapshotData) buildNodeInfoList() {
+	if data.nodeInfoList == nil {
+		data.nodeInfoList = make([]*schedulernodeinfo.NodeInfo, len(data.nodeInfoMap), len(data.nodeInfoMap)+100)
+		i := 0
+		for _, node := range data.nodeInfoMap {
+			data.nodeInfoList[i] = node
+			i++
+		}
+	}
+	if data.baseData == nil {
+		data.completeNodeInfoList = data.nodeInfoList
+		return
+	}
+	baseList := data.baseData.GetNodeInfoList()
+	totalLen := len(baseList) + len(data.nodeInfoList)
+	data.completeNodeInfoList = make([]*schedulernodeinfo.NodeInfo, totalLen, totalLen+100)
+	if len(data.deletedNodeInfos) > 0 {
+		// Oops. Costly.
+		i := 0
+		for _, bni := range baseList {
+			if data.deletedNodeInfos[bni.Node().Name] {
+				continue
+			}
+			data.completeNodeInfoList[i] = bni
+			i++
+		}
+		data.completeNodeInfoList = data.completeNodeInfoList[:i]
+	} else {
+		copy(data.completeNodeInfoList, baseList)
+	}
+	copy(data.completeNodeInfoList[len(baseList):], data.nodeInfoList)
+	return
+}
+
+func (data *internalDeltaSnapshotDataNodeLister) List() ([]*schedulernodeinfo.NodeInfo, error) {
+	return (*internalDeltaSnapshotData)(data).GetNodeInfoList(), nil
+}
+
+func (data *internalDeltaSnapshotDataNodeLister) HavePodsWithAffinityList() ([]*schedulernodeinfo.NodeInfo, error) {
+	if data.havePodsWithAffinity != nil {
+		return data.havePodsWithAffinity, nil
+	}
+
+	if data.completeNodeInfoList == nil {
+		(*internalDeltaSnapshotData)(data).buildNodeInfoList()
+	}
+
+	havePodsWithAffinityList := make([]*schedulernodeinfo.NodeInfo, len(data.completeNodeInfoList), len(data.completeNodeInfoList))
+	// i is index in *output* list.
+	i := 0
+	for _, node := range data.completeNodeInfoList {
+		if len(node.PodsWithAffinity()) > 0 {
+			havePodsWithAffinityList[i] = node
+			i++
+		}
+	}
+	data.havePodsWithAffinity = havePodsWithAffinityList[:i]
+	return data.havePodsWithAffinity, nil
+}
+
+func (data *internalDeltaSnapshotDataNodeLister) Get(nodeName string) (*schedulernodeinfo.NodeInfo, error) {
+	return (*internalDeltaSnapshotData)(data).GetNodeInfo(nodeName)
+}
+
+func (data *internalDeltaSnapshotDataPodLister) List(selector labels.Selector) ([]*apiv1.Pod, error) {
+	if data.podList == nil {
+		(*internalDeltaSnapshotData)(data).buildPodList()
+	}
+
+	if selector.Empty() {
+		// no restrictions, yay
+		return data.podList, nil
+	}
+
+	selectedPods := make([]*apiv1.Pod, len(data.podList), len(data.podList))
+	// i is index in *output* list.
+	i := 0
```

Comment:
same here.
Why not `append`?
DId I write it the same way? (maybe copied from somewhere)